File: examples/ecto/lib/shop/line_item.ex
defmodule Shop.LineItem do
  use Ecto.Schema
  @moduledoc """
  An explicit many-to-many between Order and Product, keyed by the pair.
  """

  @primary_key false

  @type t :: %__MODULE__{
          order_number: term(),
          product_id: Ecto.UUID.t(),
          quantity: integer(),
          unit_price: Decimal.t(),
          order: Shop.Order.t() | nil,
          product: Shop.Product.t() | nil
        }

  schema "line_items" do
    field(:quantity, :integer, default: 1)
    field(:unit_price, :decimal)
    field(:order_number, :id, primary_key: true)
    field(:product_id, :binary_id, primary_key: true)
    belongs_to(:order, Shop.Order, foreign_key: :order_number, define_field: false)
    belongs_to(:product, Shop.Product, foreign_key: :product_id, define_field: false, type: :binary_id)
  end
end

File: examples/eloquent/app/Models/LineItem.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * An explicit many-to-many between Order and Product, keyed by the pair.
 */
class LineItem extends Model
{
    const KEY_COLUMNS = ['order_number', 'product_id'];

    protected $table = 'line_items';

    protected $primaryKey = null;

    public $incrementing = false;

    public $timestamps = false;

    protected $fillable = [
        'order_number',
        'product_id',
        'quantity',
        'unit_price',
    ];

    protected $attributes = [
        'quantity' => 1,
    ];

    protected $casts = [
        'order_number' => 'integer',
        'quantity' => 'integer',
        'unit_price' => AsDecimal::class,
    ];

    public function getKey()
    {
        $key = [];
        foreach (static::KEY_COLUMNS as $column) {
            $key[$column] = $this->getAttribute($column);
        }

        return $key;
    }

    public function newEloquentBuilder($query)
    {
        return new CompositeKeyBuilder($query);
    }

    public function newCollection(array $models = [])
    {
        return new CompositeKeyCollection($models);
    }

    public static function destroy($ids)
    {
        $count = 0;
        foreach ((new static())->newQuery()->whereKey(func_num_args() > 1 ? func_get_args() : $ids)->get() as $model) {
            if ($model->delete()) {
                $count++;
            }
        }

        return $count;
    }

    protected function setKeysForSaveQuery($query)
    {
        foreach (static::KEY_COLUMNS as $column) {
            $query->where($column, '=', $this->original[$column] ?? $this->getAttribute($column));
        }

        return $query;
    }

    protected function setKeysForSelectQuery($query)
    {
        foreach (static::KEY_COLUMNS as $column) {
            $query->where($column, '=', $this->original[$column] ?? $this->getAttribute($column));
        }

        return $query;
    }

    public function delete()
    {
        $this->mergeAttributesFromCachedCasts();

        if (! $this->exists) {
            return null;
        }

        if ($this->fireModelEvent('deleting') === false) {
            return false;
        }

        $this->touchOwners();
        $this->performDeleteOnModel();
        $this->fireModelEvent('deleted', false);

        return true;
    }

    public function order(): BelongsTo
    {
        return $this->belongsTo(Order::class, 'order_number');
    }

    public function product(): BelongsTo
    {
        return $this->belongsTo(Product::class, 'product_id');
    }
}